doc/developer-guides/interrupt-hld.rst

```diff
@@ -0,0 +1,476 @@
+layer, it emulates virtual PIC, virtual IOAPIC and virtual LAPIC, and
+provides full APIs allowing virtual interrupt injection from emulated or
+pass-thru devices.
+
+.. figure:: images/interrupt-image3.png
+   :align: center
+   :width: 600px
+   :name: interrupt-modules-overview
+
+   ACRN Interrupt Modules Overview
+
+In the software modules view shown in :numref:`interrupt-sw-modules`,
+the ACRN hypervisor sets up the physical interrupt in its basic
+interrupt modules (e.g., IOAPIC/LAPIC/IDT).  It dispatches the interrupt
+in the hypervisor interrupt flow control layer to the corresponding
+handlers, that could be pre-defined IPI notification, timer, or runtime
+registered pass-thru devices.  The ACRN hypervisor then uses its VM
+interfaces based on vPIC, vIOAPIC, and vMSI modules, to inject the
+necessary virtual interrupt into the specific VM
+
+.. figure:: images/interrupt-image2.png
+   :align: center
+   :width: 600px
+   :name: interrupt-sw-modules
+
+   ACRN Interrupt SW Modules Overview
+
+Hypervisor Physical Interrupt Management
+****************************************
+
+The ACRN hypervisor is responsible for all the physical interrupt
+handling. All physical interrupts are first handled in VMX root-mode.
+The "external-interrupt exiting" bit in the VM-Execution controls field
+is set to support this. The ACRN hypervisor also initializes all the
+interrupt related modules such as IDT, PIC, IOAPIC, and LAPIC.
+
+Only a few physical interrupts (such as TSC-Deadline timer and IOMMU)
+are fully serviced in the hypervisor. Most interrupts come from pass-thru
+devices whose interrupt are remapped to a virtual INTx/MSI source and
+injected to the SOS or UOS, according to the pass-thru device
+configuration.
+
+The ACRN hypervisor does handle exceptions and any exception coming from
+the VMX root-mode will lead to the CPU halting. For guest exception, the
+hypervisor only traps #MC (machine check), prints a warning message, and
+injects the exception back into the guest OS.
+
+Physical Interrupt Initialization
+=================================
+
+After the ACRN hypervisor get control from the bootloader, it
+initializes all physical interrupt-related modules for all the CPUs. The
+ACRN hypervisor creates a framework to manage the physical interrupt for
+hypervisor-local devices, pass-thru devices, and IPI between CPUs.
+
+IDT
+---
+
+The ACRN hypervisor builds its native Interrupt Descriptor Table (IDT) during
+interrupt initialization. For exceptions, it links to function
+``dispatch_exception``, and for external interrupts it links to function
+``dispatch_interrupt``. Please refer to ``arch/x86/idt.S`` for more details.
+
+LAPIC
+-----
+
+The ACRN hypervisor resets LAPIC for each CPU, and provides basic APIs
+used, for example, by the local timer (TSC Deadline)
+program and IPI notification program.  These APIs include
+write_laipic_reg32, send_lapic_eoi, send_startup_ipi, and
+send_single_ipi.
+
+
+.. note::
```

Comment:
@hwang37 I put some note placeholders in where we need to have references to API docs generated from doxygen comments in the header files (where none exists today).